src/firebolt/utils/usage_tracker.py

```diff
@@ -0,0 +1,144 @@
+import inspect
+import logging
+from platform import python_version, release, system
+from sys import modules
+from typing import Dict, List, Optional, Tuple, Union
+
+from firebolt import __version__
+
+logger = logging.getLogger(__name__)
+
+
+def _os_compare(file: str, expected: str) -> bool:
+    """
+    System-independent path comparison.
+
+    Args:
+        file: file path to check against
+        expected: expected file path
+
+    Returns:
+        True if file ends with path
+    """
+    return file.endswith(expected) or file.endswith(expected.replace("/", "\\"))
+
+
+def _is_cli(func: str, file: str) -> bool:
+    return func == "create_connection" and _os_compare(file, "firebolt_cli/utils.py")
+
+
+def _is_alchemy(func: str, file: str) -> bool:
+    return func == "connect" and _os_compare(file, "sqlalchemy/engine/default.py")
+
+
+def _is_airbyte_source(func: str, file: str) -> bool:
+    return (
+        func == "establish_connection" or func == "establish_async_connection"
+    ) and _os_compare(file, "source_firebolt/source.py")
+
+
+def _is_airbyte_destination(func: str, file: str) -> bool:
+    return (
+        func == "establish_connection" or func == "establish_async_connection"
+    ) and _os_compare(file, "destination_firebolt/destination.py")
+
+
+def _is_airflow(func: str, file: str) -> bool:
+    return func == "get_conn" and _os_compare(
+        file, "firebolt_provider/hooks/firebolt.py"
+    )
+
+
+def _is_dbt(func: str, file: str) -> bool:
+    return func == "open" and _os_compare(file, "dbt/adapters/firebolt/connections.py")
+
+
+def get_sdk_properties() -> Tuple[str, str, str, str]:
+    """
+    Detect Python, OS and SDK versions.
+
+    Returns:
+        Python version, SDK version, OS name and "ciso" if imported
+    """
+    py_version = python_version()
+    sdk_version = __version__
+    os_version = f"{system()} {release()}"
+    ciso = "ciso8601" if "ciso8601" in modules.keys() else ""
+    logger.debug(
+        "Python %s detected. SDK %s OS %s %s",
+        py_version,
+        sdk_version,
+        os_version,
+        ciso,
+    )
+    return (py_version, sdk_version, os_version, ciso)
+
+
+def detect_connectors() -> Dict[str, str]:
+    """
+    Detect which connectors are running the code by parsing the stack.
+    Exceptions are ignored since this is intended for logging only.
+    """
+    connectors: Dict[str, str] = {}
+    stack = inspect.stack()
+    for f in stack:
+        try:
```

Comment:
Do you think we can replace this with a for loop?
in `_is_*` functions we basically compare `function` and `filename` to a list of strings. For version import we could've used importlib. And therefore we could basically have a map, which could be extended easily. What do you think?

Reply:
Yeah, importlib sounds like a better solution. Initially I thought we might need to use some other properties in a stack frame, but those don't look very useful.